src/plugins/intel_npu/src/al/include/sync_infer_request.hpp

```diff
@@ -163,45 +139,37 @@ class SyncInferRequest : public ov::IInferRequest {
     virtual void check_network_precision(const ov::element::Type_t precision) = 0;
 
     /**
-     * @brief Indicates a kind of provided tensor. Marks special tensors, used for internal implementation
-     */
-    enum class TensorType { InputOrOutput, Shape, State };
-
-    /**
-     * @brief Allocates a tensor on host and stores the reference inside the "_allTensors" attribute. If a buffer
-     * address is provided, then the tensor is built upon it and no additional data buffer is allocated.
-     * @param tensorName The name by which the tensor shall be identified
+     * @brief Allocates a tensor on host and stores the reference inside multiple attributes.
      * @param descriptor Tensor's metadata
-     * @param isState If true, the tensor shall also be stored inside the state variables map. In this case, adding the
-     * tensor to this structure would be required in order to correctly answer the state queries.
+     * @param isInput Determines the containers in which the newly allocated tensors will be stored.
      * @param allocator If provided, the tensor uses the custom allocator instead of using the default one.
+     * @param batchSize If provided, the value of the shape on the 0th axis is overriden with this value.
      */
-    void allocate_tensor(std::string tensorName,
-                         const IONodeDescriptor& descriptor,
-                         TensorType tensorType = TensorType::InputOrOutput,
-                         const ov::Allocator& allocator = {});
+    void allocate_tensor(const IODescriptor& descriptor,
+                         const bool isInput,
+                         const ov::Allocator& allocator = {},
+                         const std::optional<std::size_t> batchSize = std::nullopt);
 
-    // Mutable to return reference to ov::Tensor
-    mutable std::unordered_map<std::string, std::shared_ptr<ov::ITensor>> _allTensors;
-    mutable std::unordered_map<std::string, std::shared_ptr<ov::ITensor>> _shapesTensors;
-    // A copy of each tensor is needed to maintain the original L0 memory allocation in case the user provides another
-    // memory area for the tensor.
-    std::unordered_map<std::string, std::shared_ptr<ov::ITensor>> _copyAllTensors;
+    std::vector<std::shared_ptr<ov::ITensor>> _userInputTensors;
+    std::vector<std::shared_ptr<ov::ITensor>> _userOutputTensors;
 
-    std::unordered_map<std::string, std::shared_ptr<VariableState>> _variableStates;
+    std::vector<ov::SoPtr<ov::IVariableState>> _variableStates;
```

Comment:
Just for my education, why was `variableState` made as SoPtr?

Reply:
Through inheritance, the return type of `SyncInferRequest::query_state()` has to be `std::vector<ov::SoPtr<ov::IVariableState>>`. Using `SoPtr` in the referenced line allowed me to implement `query_state` in a simple manner, without doing an explicit `std::shared_ptr` -> `ov::SoPtr` for each pointer inside the vector:
```
std::vector<ov::SoPtr<ov::IVariableState>> SyncInferRequest::query_state() const {
    return _variableStates;
}
```

Not sure this is the best approach, so I'm open to suggestions, even though this may be insignificant.

On why OV has chosen this return type, I can only guess the reason is flexibility: `SoPtr` offers the possibility of storing an extra pointer towards a library, perhaps for making sure the library is still there when the destructor is called. We're not making use of this tho.